solver/src/http_solver/settlement.rs

```diff
@@ -18,92 +21,172 @@ pub struct PreparedModel {
 }
 
 pub fn convert_settlement(
-    settled: &SettledBatchAuctionModel,
-    prepared: &PreparedModel,
+    settled: SettledBatchAuctionModel,
+    prepared: PreparedModel,
 ) -> Result<Settlement> {
-    let mut settlement = Settlement::default();
-    set_orders(settled, &prepared.limit_orders, &mut settlement)?;
-    set_amms(settled, &prepared.amm_orders, &mut settlement)?;
-    set_prices(settled, &prepared.tokens, &mut settlement)?;
-    Ok(settlement)
+    let intermediate = IntermediateSettlement::new(settled, prepared)?;
+    Ok(intermediate.into_settlement())
 }
 
-fn set_orders(
-    model: &SettledBatchAuctionModel,
-    orders: &HashMap<String, LimitOrder>,
-    settlement: &mut Settlement,
-) -> Result<()> {
-    for (index, model) in model.orders.iter() {
-        let order = orders
-            .get(index.as_str())
-            .ok_or_else(|| anyhow!("invalid order {}", index))?;
-        let executed_amount = match order.kind {
-            OrderKind::Buy => model.exec_buy_amount,
-            OrderKind::Sell => model.exec_sell_amount,
-        };
-        if executed_amount.is_zero() {
-            continue;
+// An intermediate representation between SettledBatchAuctionModel and Settlement useful for doing
+// the error checking up front and then working with a more convenient representation.
+struct IntermediateSettlement {
+    executed_limit_orders: Vec<ExecutedLimitOrder>,
+    executed_amms: Vec<ExecutedAmm>,
+    prices: HashMap<H160, U256>,
+}
+
+struct ExecutedLimitOrder {
+    order: LimitOrder,
+    executed_buy_amount: U256,
+    executed_sell_amount: U256,
+}
+
+impl ExecutedLimitOrder {
+    fn executed_amount(&self) -> U256 {
+        match self.order.kind {
+            OrderKind::Buy => self.executed_buy_amount,
+            OrderKind::Sell => self.executed_sell_amount,
+        }
+    }
+}
+
+struct ExecutedAmm {
+    order: AmmOrder,
+    input: (H160, U256),
+    output: (H160, U256),
+}
+
+impl IntermediateSettlement {
+    fn new(settled: SettledBatchAuctionModel, prepared: PreparedModel) -> Result<Self> {
+        let executed_limit_orders = new_orders(prepared.limit_orders, settled.orders)?;
+        let executed_amms = new_amms(prepared.amm_orders, settled.uniswaps)?;
+        let prices = new_prices(
+            &prepared.tokens,
+            executed_limit_orders.as_slice(),
+            executed_amms.as_slice(),
+            settled.prices,
+        )?;
+        Ok(Self {
+            executed_limit_orders,
+            executed_amms,
+            prices,
+        })
+    }
+
+    fn into_settlement(self) -> Settlement {
+        let mut settlement = Settlement::default();
```

Comment:
nit: I'd slightly prefer explicitly constructing the Settlement here instead of using default. This gives us more compiler safety if we ever add a field to the settlement.